src/core/ddsc/src/dds_entity.c

```diff
@@ -1007,6 +1012,12 @@ static void pushdown_listener (dds_entity *e)
 
       ddsrt_mutex_lock (&e->m_observers_lock);
       dds_override_inherited_listener (&c->m_listener, &e->m_listener);
+      if (ddsrt_avl_is_empty(&c->m_children)) {
+        uint32_t status = ddsrt_atomic_ld32 (&c->m_status.m_status_and_mask) & SAM_STATUS_MASK;
+        if (status) {
+          dds_entity_deriver_invoke_cbs_for_pending_events(c, status);
```

Comment:
Would it not be better to also increment/decrement `m_cb_count`? I can't (yet) give an execution why it could fail without it, but at the very least there is a consistency and the fact that shaving off a pair of trivial instructions in such a rarely used piece of code isn't worth an inconsistency.

Reply:
In the revised pull-request we will increase m_cb_pending_count instead, which gives increased safety against two threads running dds_set_listener on the same entity in parallel.